Publish Artifacts CI build step does not support retries

The [Publish Artifacts CI build step](https://github.com/dotnet/sdk/blob/main/eng/pipelines/templates/jobs/vmr-build.yml#L562) will fail if rerun when a previous attempt has uploaded any artifacts.  This is because the publish build step is set to run on success or failure and that it does not support overwriting artifacts.  When the occurs, the failure will will look similar to the following:

```
##[error]Artifact CentOSStream9_Online_PreviousSourceBuiltSdk_x64_Artifacts already exists for build 2502184.
```

There are a couple options to solve this but they have some cons:

1. Only publish on success.  This would work for normal build failures but if the 1ES steps fail which are run after publish, re-run will still encounter this issue.  Furthermore, in some instance it is desirable to inspect the artifacts that were produced during a failed build.  This change would prevent that.
2. Modify the Publish Artifacts step to include the Job Attempt number.  This would mean each run would have it's artifacts uploaded to a distinct artifact name.  The downside to this, is there are some dependent jobs (e.g. the [source-build sdk content tests](https://github.com/dotnet/sdk/blob/main/src/SourceBuild/content/eng/pipelines/templates/jobs/sdk-diff-tests.yml#L63)) which download the artifacts.  They would need to account for the possibility of multiple artifacts existing.

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -159,7 +159,6 @@ jobs:
       path: $(Build.ArtifactStagingDirectory)/publishing
       artifact: $(Agent.JobName)_Artifacts
       displayName: Publish Artifacts
-      condition: always()
       sbomEnabled: true
 
     - ${{ if not(parameters.isBuiltFromVmr) }}:
@@ -548,7 +547,6 @@ jobs:
         assets/**
       TargetFolder: $(Build.ArtifactStagingDirectory)/publishing
     displayName: Copy artifacts to Artifact Staging Directory
-    condition: succeededOrFailed()
 
   # When building from source, the Private.SourceBuilt.Artifacts archive already contains the nuget packages
   - ${{ if ne(parameters.buildSourceOnly, 'true') }}:
@@ -557,11 +555,9 @@ jobs:
         SourceFolder: $(sourcesPath)/artifacts/packages
         TargetFolder: $(Build.ArtifactStagingDirectory)/publishing/packages
       displayName: Copy packages to Artifact Staging Directory
-      condition: succeededOrFailed()
 
   - ${{ if or(ne(variables['System.TeamProject'], 'internal'), eq(variables['Build.Reason'], 'PullRequest')) }}:
     - publish: $(Build.ArtifactStagingDirectory)/publishing
       artifact: $(Agent.JobName)_Artifacts
       displayName: Publish Artifacts
-      condition: succeededOrFailed()
       continueOnError: true

